tools/simple_message_relay/src/mpsc_registry.rs

```diff
@@ -0,0 +1,69 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+use tokio::sync::Mutex;
+use tokio::sync::RwLock;
+
+#[derive(Clone)]
+pub struct MpscRegistry<T: Debug> {
+    inner: Arc<RwLock<HashMap<String, (mpsc::Sender<T>, Arc<Mutex<mpsc::Receiver<T>>>)>>>,
+}
+
+impl<T: Debug> MpscRegistry<T> {
+    pub fn new() -> MpscRegistry<T> {
+        MpscRegistry {
+            inner: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub async fn subscribe(&self, user_id: &str) -> Arc<Mutex<mpsc::Receiver<T>>> {
+        let mut inner = self.inner.write().await;
+        if let Some((_, receiver)) = inner.get(user_id) {
+            return receiver.clone();
+        }
+        let (tx, rx) = mpsc::channel::<T>(100);
+        let rx = Arc::new(Mutex::new(rx));
+        inner.insert(user_id.to_owned(), (tx, rx.clone()));
+        rx
+    }
+
+    // todo: don't unwrap, return result
+    pub async fn send_msg(&self, user_id: &str, message: T) {
+        let inner = self.inner.read().await;
+        if let Some((sender, _)) = inner.get(user_id) {
+            sender.send(message).await.unwrap()
+        } else {
+            warn!("User {} not found.", user_id);
+        }
+    }
+}
```

Comment:
I admit that I didn't checkout the PR to look through the code in an IDE but I feel like this is not a good way to go about messages.

The channels are fine. The problem I see is the web server and this shared ownership registry with interior mutability.

Unless I'm mistaken, the web server is there to allow the reuse of the HttpClient for sending messages. My curiosity is: why not use a different implementor of the `Transport` type that relies entirely on channels, thus removing the need for a web server?

Secondly, channels could be distributed as needed to their agents based on the connections between them. This would remove the need to share the whole set of senders/receivers with all the agents, but rather allow each agent to have a channel per connection which it exclusively owns.

It's a big change and I didn't think of all the implementation details so feel free to disregard it, but that might speed things up even more and make them more idiomatic.

Reply:
It's a fair point. Arguably for demo purposes you might want to be closer to real world, eg. using http mediator; but on the other hand, the existence of the http mediator was largely abstracted away in this demo anyway.

I'll loo into this